### ISSUE TITLE
HoloLens graphics crash

### ISSUE THREAD

**Issue by maluoi** [StereoKit/StereoKit#528]
## Description

From Christoph on Discord:

Recently encountered this weird error when starting the app. Any idea what could cause this? It happens very rarely but always appears right after starting the app and when SK is already running. It seems like the frames don't get cleared but just draw on top of each other. We are running some asynchronous code (specifically request permissions https://learn.microsoft.com/hu-hu/uwp/api/windows.security.authorization.appcapabilityaccess.appcapability.requestaccessforcapabilitiesasync?view=winrt-22621&viewFallbackFrom=winrt-14393) between SK.Initialize and SK.Run, could this have anything to do with it? Possibly looks related to this issue: https://github.com/StereoKit/StereoKit/issues/361

![bug_multiple](https://user-images.githubusercontent.com/8484884/201822146-183a0ace-e517-41a9-9903-57ba4451bca7.jpg)

## Platform / Environment

UWP/HoloLens 2

## Logs or exception details

Here's some abbreviated logs
```
[SK info] [RequestAccess::GetNativeCapabilities] Requesting eye gaze permission.
[SK info] [RequestAccess::GetNativeCapabilities] Requesting webcam permission.
[SK info] [RequestAccess::GetNativeCapabilities] Requesting microphone permission.
[SK info] [RequestAccess::GetInternalCapabilities] Requesting qr code permission.
[SK info] [RequestAccess::GetInternalCapabilities] Requesting research mode permission.
[SK info] [RequestAccess::GetCapabilites] All native capabilities passed: True.
[SK info] [RequestAccess::GetCapabilites] All internal capabilities passed: True.
[SK error] sk_gpu: CreateBuffer failed!
[SK error] sk_gpu: Create texture error!
[SK error] mesh_set_inds: Failed to create dynamic index buffer
[SK error] sk_gpu: Create Shader Resource View error!
[SK error] sk_gpu: CreateBuffer failed!
[SK error] mesh_set_verts: Failed to create dynamic vertex buffer
[SK error] sk_gpu: CreateBuffer failed!
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed!
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed!
[SK error] mesh_set_verts: Failed to create vertex buffer
```

And pseudocode for the application flow
```csharp
static void Main(string[] args)
{
    SKSettings settings = new()
    {
        appName = applicationName,
        assetsFolder = "Assets",
    };
    Task someTask = doSomeTask();

    if (!SK.Initialize(settings))
        Environment.Exit(1);
    
    Action onComplete = async () =>
    {
        await someTask;
        SK.AddStepper<Menu>();
    };

    RequestAccess access = new(onComplete);
    SK.AddStepper(access);

    SK.Run();
}

class RequestAccess : IStepper
{
    private Action onComplete;
    RequestAccess(Action onComplete)
    {
        this.onComplete ??= onComplete;
    }

    public void Step()
    {
        switch (State)
        {
            case MenuState.RequestCapabilities:
                State = MenuState.RequestingCapabilities;
                // Run the async task without await so that it runs on another worker thread, separate from
                // the SK initialization/stepper thread (which has async Asset creation in https://github.com/StereoKit/StereoKit/issues/361)
                _ = this.GetCapabilites(); // run async task
                break;
            case MenuState.RequestingCapabilities:
                break; // do nothing
            case MenuState.PermissionsMenu:
                this.PermissionsMenu(); // show missing permissions UI
                break;
            case MenuState.Shutdown:
                SK.RemoveStepper(this);
                this.onComplete?.Invoke(); // add other steppers
                break;
        }
    }
}

class Menu : IStepper
{
    bool Initialize()
    {
        someSprite = Sprite.FromFile("some.png", SpriteType.Single);
        return true;
    }
}
```

**Comment by austinbhale** [comment]
Is it too late for https://github.com/StereoKit/StereoKit/commit/fcd2a949641d5df3ce357bac325fd6693bef1361 to be in the newest preview build? :D 

No worries if not, we can reference the dev branch instead for testing

**Comment by maluoi** [comment]
Unfortunately, it is. I've got some logging additions incoming here as well, and I'll pass along a .nupkg on Discord when that's done! If that _does_ solve things, I can probably push out another preview build shortly after :)

**Comment by austinbhale** [comment]
Cool, thank you! I’d noticed that Step had been called a few times before Initialize completed in another instance before this. This could help remove some null checks!

Also, a question on the thread safety of async asset creation (materials/meshes) on the gpu similar to #361: does SK ensure that they are being called from the proper threads?

For example, in this case, is it generally safe to load a sprite from file on _any_ thread, whether that be via a worker thread, task pool, CoreDispatcher UI, main, etc? Or does SK require that it initially be called from a proper SK thread?

This is slightly different than #512, since there is no async code between Initialize and Step, instead it I’m curious about the thread that calls the Stepper.

> "When you start an async method, the current thread is used for the async call, and the rest of the code continues on a new thread"

If you call an async method from the Step method, does this mean that the game thread can go into the async method?

**Comment by paulmelis** [comment]
I've seen this once, but only very recently. I'm not sure it happened only after updating to the latest preview, but perhaps something relevant changed recently in SK's codebase?

**Comment by maluoi** [comment]
Okay, that's really interesting! `0x887A0005` is a `DXGI_ERROR_DEVICE_REMOVED` message. This seems kinda extreme, and I'm not certain what exactly would cause this to happen. My only current thought is that this is still thread related, where maybe it's somehow now sending graphics commands from a different thread?

[This document](https://learn.microsoft.com/en-us/windows/uwp/gaming/handling-device-lost-scenarios) has a bunch of information about why this message might pop up. I can reproduce the same sort of behavior with the `dxcap -forcetdr` command mentioned there, but that unfortunately doesn't give me much insight either.

I'll add some additional logs to get the "device removed reason", this might help pin it down a little more. I'll also mess a bit more with the sample code from before, see if I can get it to reproduce this condition specifically.

**Comment by austinbhale** [comment]
Still on the hunt for pinpointing the exact behavior, but I've gotten it to crash in this way several times today by messing around with different locations on startup, going to System > Holograms > Remove holograms, and covering the tracking sensors for SLAM. Below is the updated log output from the most recent crash. I made sure to install the newest NuGet package you sent, but it doesn't seem like your newest logs were found, so maybe it's occurring in a different location? Also, in this abbreviated log, I found it interesting that it is successful for 30+ frames with the steppers, shows the errors, and then continues to move through the steppers. (The numbers just indicate random lines of code from the beginning to end of the same Step method)

```
[SK diagnostic] sk_gpu: Using Direct3D 11: Qualcomm(R) Adreno(TM) 630 GPU
[SK diagnostic] Starting mixed reality mode
[SK diagnostic] available: XR_KHR_D3D12_enable
[SK diagnostic] available: XR_KHR_composition_layer_color_scale_bias
[SK diagnostic] available: XR_EXT_win32_appcontainer_compatible
[SK diagnostic] available: XR_EXT_conformance_automation
[SK diagnostic] available: XR_MSFT_composition_layer_reprojection
[SK diagnostic] available: XR_MSFT_spatial_anchor_persistence
[SK diagnostic] available: XR_MSFT_scene_understanding_serialization
[SK diagnostic] available: XR_EXT_hand_joints_motion_range
[SK diagnostic] available: XR_MSFT_holographic_window_attachment
[SK diagnostic] available: XR_EXT_debug_utils
[SK diagnostic] REQUESTED: XR_KHR_win32_convert_performance_counter_time
[SK diagnostic] REQUESTED: XR_KHR_D3D11_enable
[SK diagnostic] REQUESTED: XR_KHR_composition_layer_depth
[SK diagnostic] REQUESTED: XR_MSFT_unbounded_reference_space
[SK diagnostic] REQUESTED: XR_MSFT_spatial_anchor
[SK diagnostic] REQUESTED: XR_MSFT_perception_anchor_interop
[SK diagnostic] REQUESTED: XR_MSFT_spatial_graph_bridge
[SK diagnostic] REQUESTED: XR_MSFT_scene_understanding
[SK diagnostic] REQUESTED: XR_EXT_eye_gaze_interaction
[SK diagnostic] REQUESTED: XR_MSFT_hand_interaction
[SK diagnostic] REQUESTED: XR_EXT_hand_tracking
[SK diagnostic] REQUESTED: XR_MSFT_hand_tracking_mesh
[SK diagnostic] REQUESTED: XR_MSFT_secondary_view_configuration
[SK diagnostic] REQUESTED: XR_MSFT_first_person_observer
[SK diagnostic] Using system: Windows Mixed Reality
[SK diagnostic] Platform supports single-pass rendering
[SK diagnostic] OpenXR articulated hands ext enabled!
[SK diagnostic] OpenXR hand mesh ext enabled!
[SK diagnostic] OpenXR depth LSR ext enabled!
[SK diagnostic] OpenXR gaze ext enabled!
[SK diagnostic] OpenXR spatial bridge ext enabled!
[SK diagnostic] OpenXR perception anchor interop ext enabled!
[SK diagnostic] OpenXR world occlusion enabled! (Scene Understanding)
[SK diagnostic] OpenXR world raycast enabled! (Scene Understanding)
[SK diagnostic] Creating view: PrimaryStereo color:bgra32_sRGB depth:depth16 blend:Additive
[SK diagnostic] Setting view: PrimaryStereo to 1440x936
[SK diagnostic] Creating view: SecondaryMonoFPO_Msft color:bgra32_sRGB depth:depth16 blend:Alpha Blend
[SK diagnostic] Initializing Audio
[SK info] Using audio backend: ISAC
[SK diagnostic] Initializing Assets
[SK diagnostic] Initializing Defaults
[SK diagnostic] Initializing World
[SK diagnostic] Initializing Sprites
[SK diagnostic] Initializing Lines
[SK diagnostic] Initializing UI
[SK diagnostic] Initializing Renderer
[SK diagnostic] Initializing Physics
[SK diagnostic] Initializing Input
[SK info] Initialization successful
[SK diagnostic] OpenXR session begin.
[SK info] [RequestAccess::GetNativeCapabilities] Requesting microphone permission.
[SK info] [RequestAccess::GetNativeCapabilities] Requesting webcam permission.
[SK info] [RequestAccess::GetNativeCapabilities] Requesting eye gaze permission.
[SK info] [RequestAccess::GetInternalCapabilities] Requesting qr code permission.
[SK info] [RequestAccess::GetInternalCapabilities] Requesting research mode permission.
[SK info] [RequestAccess::GetCapabilites] All native capabilities passed: True.
[SK info] [RequestAccess::GetCapabilites] All internal capabilities passed: True.
[SK info] [RequestAccess::Shutdown] Calling access granted.
[SK info] [Program::Main] ONCOMPLETE BEFORE
[SK info] [Program::RegisterISteppers] DEBUGWINDOW
[SK info] [Program::RegisterISteppers] HANDMENU
[SK info] [Program::RegisterISteppers] NAKAMIRINPUT
[SK info] [NakamirInput::Initialize] 1
[SK info] [NakamirInput::Initialize] 2
[SK info] [Program::RegisterISteppers] ADDED STEPPERS
[SK info] [Program::Main] ONCOMPLETE AFTER
[SK info] [TimedText::Show] 1
[SK info] [TimedText::Show] 2
[SK info] [TimedText::Start] 1
[SK info] [TimedText::Start] 2
[SK info] [TimedText::Start] 3
[SK info] [TimedText::Show] 3
[SK info] [RequestAccess::Shutdown] Called access granted.
[SK info] [HandMenu::DrawHandMenu] 1
[SK info] [HandMenu::DrawHandMenu] 2
[SK info] [HandMenu::DrawHandMenu] 3
...
[SK info] [DebugLogWindow::ShowLogWindow] 1
[SK info] [DebugLogWindow::ShowLogWindow] 2
[SK diagnostic] Switched right controller profile to microsoft/hand_interaction
[SK info] [DebugLogWindow::ShowLogWindow] 1
[SK info] [DebugLogWindow::ShowLogWindow] 2
[SK info] [DebugLogWindow::ShowLogWindow] 3
[SK info] [DebugLogWindow::ShowLogWindow] 4
[SK info] [HandMenu::DrawHandMenu] 1
[SK info] [HandMenu::DrawHandMenu] 2
[SK info] [HandMenu::DrawHandMenu] 3
[SK info] [NakamirInput::Step] 1
[SK info] [NakamirInput::Step] 3
[SK info] [NakamirInput::Step] 4
[SK info] [NakamirInput::ToSKEyes] 1
[SK info] [NakamirInput::ToSKEyes] 2
[SK info] [NakamirInput::ToSKEyes] 3
[SK info] [NakamirInput::Step] 5
[SK info] [TimedText::Step] 1
[SK info] [TimedText::Step] 4
[SK info] [TimedText::Step] 5
[SK info] [DebugLogWindow::ShowLogWindow] 1
[SK info] [DebugLogWindow::ShowLogWindow] 2
[SK info] [DebugLogWindow::ShowLogWindow] 3
[SK info] [DebugLogWindow::ShowLogWindow] 4
[SK info] [HandMenu::DrawHandMenu] 1
[SK info] [HandMenu::DrawHandMenu] 2
[SK info] [HandMenu::DrawHandMenu] 3
[SK info] [NakamirInput::Step] 1
[SK info] [NakamirInput::Step] 3
[SK info] [NakamirInput::Step] 4
[SK info] [NakamirInput::ToSKEyes] 1
[SK info] [NakamirInput::ToSKEyes] 2
[SK info] [NakamirInput::ToSKEyes] 3
[SK info] [NakamirInput::Step] 5
[SK info] [TimedText::Step] 1
[SK info] [TimedText::Step] 4
[SK info] [TimedText::Step] 5
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_inds: Failed to create dynamic index buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create dynamic vertex buffer
[SK info] [DebugLogWindow::ShowLogWindow] 1
[SK info] [DebugLogWindow::ShowLogWindow] 2
[SK info] [DebugLogWindow::ShowLogWindow] 3
[SK info] [DebugLogWindow::ShowLogWindow] 4
[SK info] [HandMenu::DrawHandMenu] 1
[SK info] [HandMenu::DrawHandMenu] 2
[SK info] [HandMenu::DrawHandMenu] 3
[SK info] [NakamirInput::Step] 1
[SK info] [NakamirInput::Step] 3
[SK info] [NakamirInput::Step] 4
[SK info] [NakamirInput::ToSKEyes] 1
[SK info] [NakamirInput::ToSKEyes] 2
[SK info] [NakamirInput::ToSKEyes] 3
[SK info] [NakamirInput::Step] 5
[SK info] [TimedText::Step] 1
[SK info] [TimedText::Step] 4
[SK info] [TimedText::Step] 5
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_inds: Failed to create index buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK info] [DebugLogWindow::ShowLogWindow] 1
[SK info] [DebugLogWindow::ShowLogWindow] 2
[SK info] [DebugLogWindow::ShowLogWindow] 3
[SK info] [DebugLogWindow::ShowLogWindow] 4
[SK info] [HandMenu::DrawHandMenu] 1
[SK info] [HandMenu::DrawHandMenu] 2
[SK info] [HandMenu::DrawHandMenu] 3
[SK info] [NakamirInput::Step] 1
[SK info] [NakamirInput::Step] 3
[SK info] [NakamirInput::Step] 4
[SK info] [NakamirInput::ToSKEyes] 1
[SK info] [NakamirInput::ToSKEyes] 2
[SK info] [NakamirInput::ToSKEyes] 3
[SK info] [NakamirInput::Step] 5
[SK info] [TimedText::Step] 1
[SK info] [TimedText::Step] 4
[SK info] [TimedText::Step] 5
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK info] [DebugLogWindow::ShowLogWindow] 1
[SK info] [DebugLogWindow::ShowLogWindow] 2
[SK info] [DebugLogWindow::ShowLogWindow] 3
[SK info] [DebugLogWindow::ShowLogWindow] 4
[SK info] [HandMenu::DrawHandMenu] 1
[SK info] [HandMenu::DrawHandMenu] 2
[SK info] [HandMenu::DrawHandMenu] 3
[SK info] [NakamirInput::Step] 1
[SK info] [NakamirInput::Step] 3
[SK info] [NakamirInput::Step] 4
[SK info] [NakamirInput::ToSKEyes] 1
[SK info] [NakamirInput::ToSKEyes] 2
[SK info] [NakamirInput::ToSKEyes] 3
[SK info] [NakamirInput::Step] 5
[SK info] [TimedText::Step] 1
[SK info] [TimedText::Step] 4
[SK info] [TimedText::Step] 5
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_inds: Failed to create index buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK info] [DebugLogWindow::ShowLogWindow] 1
[SK info] [DebugLogWindow::ShowLogWindow] 2
[SK info] [DebugLogWindow::ShowLogWindow] 3
[SK info] [DebugLogWindow::ShowLogWindow] 4
[SK info] [HandMenu::DrawHandMenu] 1
[SK info] [HandMenu::DrawHandMenu] 2
[SK info] [HandMenu::DrawHandMenu] 3
[SK info] [NakamirInput::Step] 1
[SK info] [NakamirInput::Step] 3
[SK info] [NakamirInput::Step] 4
[SK info] [NakamirInput::ToSKEyes] 1
[SK info] [NakamirInput::ToSKEyes] 2
[SK info] [NakamirInput::ToSKEyes] 3
[SK info] [NakamirInput::Step] 5
[SK info] [TimedText::Step] 1
[SK info] [TimedText::Step] 4
[SK info] [TimedText::Step] 5
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_inds: Failed to create index buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK info] [DebugLogWindow::ShowLogWindow] 1
[SK info] [DebugLogWindow::ShowLogWindow] 2
[SK info] [DebugLogWindow::ShowLogWindow] 3
[SK info] [DebugLogWindow::ShowLogWindow] 4
[SK info] [HandMenu::DrawHandMenu] 1
[SK info] [HandMenu::DrawHandMenu] 2
[SK info] [HandMenu::DrawHandMenu] 3
[SK info] [NakamirInput::Step] 1
[SK info] [NakamirInput::Step] 3
[SK info] [NakamirInput::Step] 4
[SK info] [NakamirInput::ToSKEyes] 1
[SK info] [NakamirInput::ToSKEyes] 2
[SK info] [NakamirInput::ToSKEyes] 3
[SK info] [NakamirInput::Step] 5
[SK info] [TimedText::Step] 1
[SK info] [TimedText::Step] 4
[SK info] [TimedText::Step] 5
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_inds: Failed to create index buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK info] [DebugLogWindow::ShowLogWindow] 1
...
```

**Comment by maluoi** [comment]
Is it possible it's still using the previous build? When sharing the same version number like those two builds did, you will need to manually delete the NuGet folder ( `C:\Users\[username]\.nuget\packages\stereokit\0.3.7-preview.7` ) to force it to re-install the new package.

The SLAM blocking thing is kinda weird! I'll have to give this a try. When I tried the `dxcap -forcetdr` command, SK did continue to run there as well. In theory some types of device removed events _can_ be recovered from, so I wouldn't necessarily expect it to crash, it's just a _lot_ of work for all parties involved.

**Comment by cwule** [comment]
It feels like the error now happens less, but unfortunately it still appears occasionally during startup, so this may be a placebo effect :)

**Comment by HarrisonBT** [comment]
I managed to stop this error from happening by putting a wait in before the run.Step()
I guess this gives gpu time to load everything before anything is drawn.

        app.Init();

        Task.Delay(15000).Wait(); // time to load everything

        // Now loop until finished, and then shut down
        SK.Run(app.Step);

**Comment by austinbhale** [comment]
@HarrisonBT were you able to get the crash consistently?

**Comment by HarrisonBT** [comment]
> @HarrisonBT were you able to get the crash consistently?

Pretty much if I draw() something before it had loaded. (our models are quite large), 
I found if I load a model and draw it straight away I get this error.

If I load the model, and then wait a few seconds before drawing it, It's fine.

**Comment by austinbhale** [comment]
Hey @maluoi, I notice that the steppers functionality is not thread safe since it uses a generic list. I think incorporating a lightweight lock or the near "lockless" Concurrent class (e.g., maybe a ConcurrentDictionary, though ordering would be tricky) whenever the steppers collection changes would give lots of flexibility to devs by not worrying if adding/removing a stepper from a non-"game thread" might cause problems. I could rough out some solution too, ideally one that doesn't hang on the Step() method for too long. Curious what you think, thanks!

**Comment by maluoi** [comment]
@austinbhale I'm spinning this bit off into a separate issue, unless you think that's more directly related here! I did put together a small test app to check out the idea, and yeah, it definitely can cause problems. But we can continue the discussion over there!

**Comment by austinbhale** [comment]
Just got this error again on the develop branch, when running the demos. I may have pressed two demos at once, according to the log. The device removed text was present:
```
[SK info] Starting Scene: DemoUI
[SK info] Starting Scene: DemoWorldAnchor
The thread 0x19fc has exited with code 0 (0x0).
D3D11: Removing Device.
The thread 0x1794 has exited with code 0 (0x0).
The thread 0x198c has exited with code 0 (0x0).
[SK info] App focus changed to: Background
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create dynamic vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create dynamic vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
The thread 0x1870 has exited with code 0 (0x0).
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
[SK error] sk_gpu: CreateBuffer failed: 0x887A0005
[SK error] mesh_set_verts: Failed to create vertex buffer
```

**Comment by maluoi** [comment]
`D3D11: Removing Device.` and `App focus changed to: Background` look like new messages here! Were those happening at all before..? App going to background seems kinda interesting in this case.

I'm going to switch over to HL2/UWP as my main dev device for a bit here, so hopefully I'll run into it as well.

**Comment by maluoi** [comment]
I've been able to repro this fairly consistently with flipping between large photogrammetry on HoloLens 2 right now. I do get some additional debug info in my current setup, and I was able to get this message:

```
D3D11: Removing Device.
D3D11 ERROR: ID3D11Device::RemoveDevice: Device removal has been triggered for the following reason (DXGI_ERROR_DEVICE_HUNG: The Device took an unreasonable amount of time to execute its commands, or the hardware crashed/hung. As a result, the TDR (Timeout Detection and Recovery) mechanism has been triggered. The current Device Context was executing commands when the hang occurred. The application may want to respawn and fallback to less aggressive use of the display hardware). [ EXECUTION ERROR #378: DEVICE_REMOVAL_PROCESS_AT_FAULT]
```

That definitely suggests that the photogrammetry assets could just be taking too long, so the GPU panics and gives up. _Beyond_ that, I also get this message a few lines earlier! Proximity is close enough that I suspect it's probably the root cause.

```
D3D11 INFO: ID3D11Device::CreateTexture2D: Note that the resource allocation (357913952 bytes plus overhead) would use more than 128 MB of application usable memory. This is fine; D3D allows attempts to make allocations above 128 MB in the event that they may happen to succeed, however this usage is subject to hardware specific failure.  D3D11 only guarantees that allocations within 128 MB are supported by all D3D11 hardware.  Here, failure may only happen if the system runs out of resources. Allocations above 128 MB may fail for a couple of reasons, not only because the system is overextended, but also if the particular hardware being used does not support it.  There is intentionally no supported way to report individual hardware limits on allocation sizes above 128 MB. [ STATE_CREATION INFO #107: CREATETEXTURE2D_LARGEALLOCATION]
```

All of the photogrammetry assets I was testing with here were using 8k textures, which weigh in at over 300mb! Sometimes it crashes on the first model, sometimes the third or fourth, but it seems like this may be pretty shaky ground to begin with. I'll check some more with 4k textures, and I'd love to know if anyone here has been using 5k or more.

If the 128mb thing is a reliability limit like it seems, then I'd probably recommend capping your textures at 4k or so (technically 33,554,432 pixels, or 5792x5792) and see if that helps. Texture compression will likely help a _lot_ here as well, but that's work I still need to do for StereoKit myself.

**Comment by maluoi** [comment]
Okay, I was perhaps a little too excited! I'm still getting this with 4k textures too.

**Comment by cwule** [comment]
Glad the bug still showed up, because we never use 8k textures :)
Our largest texture size when the app is already running is 896x504 (the camera image) and the bug sometimes appears when these are the only textures appearing. We will test whether we still observe it when we don't use the camera texture at all (which is unfortunately just an option for debugging, not for the actual app).

**Comment by cwule** [comment]
Tested the preview8 for a couple of hours today on the Hololens 2. I was already very excited, because the bug didn't appear for a very long time (definitely less than before), but finally after running it for about two hours it appeared again. Our largest texture that time was the camera image with 896x504 and with a framerate of 15fps. We had some problem with the logs unfortunately, so no new logs yet, will provide them if I can recreate it again.

**Comment by austinbhale** [comment]
Thanks for sending the latest preview build in Discord! I received a `D3D11 CORRUPTION` error when calling `Tex.SetColors` from a separate thread (not SK game thread). I ensured that the calling of this `SetColors` method is synchronous, so there aren't any race conditions in our application. Here are the relevant details:

```
D3D11 CORRUPTION: ID3D11DeviceContext::RSSetViewports: Two threads were found to be executing functions associated with the same Device[Context] at the same time. This will cause corruption of memory. Appropriate thread synchronization needs to occur external to the Direct3D API (or through the ID3D10Multithread interface). 1008 and 7936 are the implicated thread ids. [ MISCELLANEOUS CORRUPTION #28: CORRUPTED_MULTITHREADING]
Exception thrown at 0x763F406F (KernelBase.dll) in NARA.exe: 0x0000087D (parameters: 0x00000000, 0x007BD7CC, 0x007BCBF8).
```

**Comment by austinbhale** [comment]
The texture mentioned above has bit flags for `TexType.Dynamic | TexType.ImageNomips` so it says the data is corrupt at the following `Map` calls:
https://github.com/StereoKit/StereoKit/blob/7c9a711e37fafc079d124dbf5405edb8ee67f597/StereoKitC/libraries/sk_gpu.h#L1109 https://github.com/StereoKit/StereoKit/blob/7c9a711e37fafc079d124dbf5405edb8ee67f597/StereoKitC/libraries/sk_gpu.h#L2220 
There's a comment that D3D works with multiple threads https://github.com/StereoKit/StereoKit/blob/7c9a711e37fafc079d124dbf5405edb8ee67f597/StereoKitC/asset_types/texture.cpp#L855 but the error message below seems to show otherwise?
```
D3D11 CORRUPTION: ID3D11DeviceContext::Map: Two threads were found to be executing functions associated with the same Device[Context] at the same time. This will cause corruption of memory. Appropriate thread synchronization needs to occur external to the Direct3D API (or through the ID3D10Multithread interface). 3484 and 9200 are the implicated thread ids. [ MISCELLANEOUS CORRUPTION #28: CORRUPTED_MULTITHREADING]
```

**Comment by maluoi** [comment]
Yeah, these errors look like something I can dig into, I think this is what I was hoping to get from that build!

This is actually the first time I've ever seen D3D unhappy with multithreading! SK's asset thread has been tossing stuff at the GPU thread for a while now, and I haven't ever seen this issue on PC. I wonder if it's a difference in drivers on HoloLens?

Either way, I'll try saturating this path and see if I can repro it. The debug message alone gives me a fair bit I can read up on, thanks a bunch for this!

**Comment by maluoi** [comment]
Okay, `d3d_device` is always threadsafe, but `d3d_context` may need a little extra work to be completely threadsafe!That means creating a new texture is fine, but `Map`ping a dynamic one is not.

As a workaround, you may be able to just create a new `Tex` every time? Wasteful for sure, but if it's not too wasteful, that should bypass some of what's going on here.

**Comment by maluoi** [comment]
Alright, I was reliably replicating the exact same error, and I've now also got a fix for that! I'd still like some time to clean it up and think about some alternative ways of doing it, but it does work now. Let me know how much of a rush you're in, I can get it to you sooner rather than later if you need it.

**Comment by maluoi** [comment]
Forgot to tag the commits with this issue, but a fix for the multithreaded SetColors call is here: b39511c2ee0da0cf4718217cb830d95fa3e70033, and there's now a test for the situation over here as well: 0a54aae38b485c8faad425a49d21bcd48c19df7d

**Comment by austinbhale** [comment]
Thanks my man! You definitely came in clutch with this fix :) so far we've seen no multi-window bugs. I'll update you if we see any and will run the test case you've written if we can get it consistently from some saved sequence.

Looking at your commit,, with this safer approach to multithreading (single lock) which has the main context wait for all deferred tasks to finish, would there be potential bottlenecks if we have many Map/Unmap calls? It seems that the deferred and main contexts maintain order with who grabs the lock next. Hypothetically, do many Map/Unmap calls in the deferred context directly impact frame rate?

**Comment by maluoi** [comment]
The big thing this fix does is schedule the map/unmap calls until right before normal rendering! There is overhead due to the deferred context allocating and storing memory for those map/unmap calls, which I think is probably the biggest catch to it. One thing that could require a bit of watching, if the frame never actually gets drawn, but SetColors continues to be called, then the deferred context will continue to grow in size. You can see this if you run the test scenario in flatscreen and drag the window around, thereby preventing it from presenting regularly, but I expect this is probably not common in real world scenarios. The test runs at an uncapped speed independent of any framerate.

I've also considered doing a texture double-buffer for dynamic textures? Where one is active for drawing, and the other can be written to while the other busy. That's theoretically safe even with the immediate context! However, this becomes a bit more complicated architecturally considering that multiple writes happening per-frame would prevent simple swapping. If the deferred context becomes problematic, I might give this approach a try.

I didn't measure the performance impact of it, but it was working pretty nicely for me. This might be something for me to investigate a little deeper in the future. I'd love to know if it becomes problematic for you!

**Comment by austinbhale** [comment]
I tested the performance of `Tex.SetColors` between `preview.7` vs. https://github.com/StereoKit/StereoKit/commit/b39511c2ee0da0cf4718217cb830d95fa3e70033 on our use case of playing a 20-second video stream (896x504 @ 15 FPS; NV-12 encoded) using a dynamic NoMips texture. I didn't grab the times of the overall impact on the game thread or the steppers, just Tex.SetColors from a multithreaded POV. I ran it three times for each since the timing can vary, which are represented by different colors. The y-axis is in milliseconds.

![Preview 7 - Tex SetColors](https://user-images.githubusercontent.com/32556981/210841220-90420cdd-4eea-4f0a-885b-d533dab68bcc.png)

![b39511c - Tex SetColors](https://user-images.githubusercontent.com/32556981/210841386-f19a799e-fbb9-4312-bd8e-c46e5706b266.png)

Though there's only a ~2 ms average difference between the two, I am consistently noticing larger and more frequent spikes in this latest commit than the previous (unsafe) single context way. Do you think these spikes are due to the increased overhead with our fairly frequent calls to `Tex.SetColor` or could it be that the immediate context is busy so it waits a bit?

We use a similar double buffering technique where we have a couple memory pools and multiple pointers to ensure that there is always one pointing to the most recently filled buffer, while the other continues to fill. I'm curious if double buffering for dynamic textures would smoothen out the spikes since it shouldn't hang any longer than the normal Map/Unmap calls?

**Comment by austinbhale** [comment]
We haven't received anymore of these crashes after a week of testing! So do you think this issue should be good to resolve?

BTW, I think your work on texture compression would probably make the performance notes above obsolete :)

**Comment by maluoi** [comment]
Super glad to hear that seems to have solved it! I'll mark this as done when the next preview build goes out :)

I do have a guess about the performance here, I suspect it may be a bit more about the timing of the deferred `Map` calls? Right now they're deferred until right before drawing the frame, which could be more consistently delaying the GPU as it uploads the data, and other calls are waiting on that. If I was able to defer it to the start of the current frame, or even the start of the next frame, the GPU might have a fair more time to complete the upload before drawing everything else. That's my current thought anyway!

**Comment by maluoi** [comment]
This is now in v0.3.7-preview.8!

**Comment by HarrisonBT** [comment]
I still get with preview8
(sometimes)

09:39:08:989	D3D11: Removing Device.
09:39:08:989	[SK error] [sk_gpu] CreateBuffer failed: 0x887A0005
09:39:08:989	[SK error] mesh_set_verts: Failed to create vertex buffer
09:39:08:989	[SK error] [sk_gpu] CreateBuffer failed: 0x887A0005
09:39:08:989	[SK error] mesh_set_inds: Failed to create index buffer
09:39:08:989	[SK error] [sk_gpu] CreateBuffer failed: 0x887A0005
09:39:08:989	[SK error] mesh_set_inds: Failed to create dynamic index buffer
09:39:08:989	The thread 0xa14 has exited with code 0 (0x0).
09:39:08:989	[SK error] [sk_gpu] CreateBuffer failed: 0x887A0005
09:39:08:989	[SK error] mesh_set_verts: Failed to create dynamic vertex buffer

**Comment by maluoi** [comment]
Are you doing any multithreaded graphics operations that look suspicious here @HarrisonBT? Any ideas what the root cause of yours might be? Would love a bit more info about how I could reproduce your issue :)

The `D3D11: Removing Device` seems to be a pretty generic general fail message for when the GPU runs into a problem downstream of the call and can't do something.

**Comment by HarrisonBT** [comment]
I'm loading all my models in a stepper at application start, calling an async Task function
'private async Task loadModels()' within my stepper.

I get the error less when I put an 'await Task.Delay()' in between loading each model.
I'm also sure to not draw any models until they have all loaded

**Comment by HarrisonBT** [comment]
I also sometimes get artifacts in my 3d models.
Does this error have something to do with ?

 Too many texture coordinate channels! Only one supported, the rest will be ignored.
